File: examples/examples/udf_example.rs
use std::any::Any;
use std::sync::Arc;
use std::time::Duration;

use datafusion::common::cast::as_float64_array;
use datafusion::functions_aggregate::average::avg;
use datafusion::functions_aggregate::count::count;
use datafusion::functions_aggregate::expr_fn::{max, min};
use datafusion::logical_expr::sort_properties::{ExprProperties, SortProperties};
use datafusion::logical_expr::Volatility;
use datafusion::logical_expr::{col, lit};
use datafusion::logical_expr::{ColumnarValue, ScalarUDF, ScalarUDFImpl, Signature};

use denormalized::datasource::kafka::{ConnectionOpts, KafkaTopicBuilder};
use denormalized::prelude::*;

use arrow::array::{ArrayRef, Float64Array};
use arrow::datatypes::DataType;

use denormalized_examples::get_sample_json;

#[tokio::main]
async fn main() -> Result<()> {
    env_logger::builder()
        .filter_level(log::LevelFilter::Info)
        .init();

    let sample_event = get_sample_json();

    let bootstrap_servers = String::from("localhost:9092");

    let ctx = Context::new()?;
    let mut topic_builder = KafkaTopicBuilder::new(bootstrap_servers.clone());

    let source_topic = topic_builder
        .with_topic(String::from("temperature"))
        .infer_schema_from_json(sample_event.as_str())?
        .with_encoding("json")?
        .with_timestamp(String::from("occurred_at_ms"), TimestampUnit::Int64Millis)
        .build_reader(ConnectionOpts::from([
            ("auto.offset.reset".to_string(), "latest".to_string()),
            ("group.id".to_string(), "sample_pipeline".to_string()),
        ]))
        .await?;

    let sample_udf = ScalarUDF::from(SampleUdf::new());

    let ds = ctx
        .from_topic(source_topic)
        .await?
        .window(
            vec![col("sensor_name")],
            vec![
                count(col("reading")).alias("count"),
                min(col("reading")).alias("min"),
                max(col("reading")).alias("max"),
                avg(col("reading")).alias("average"),
            ],
            Duration::from_millis(1_000),
            None,
        )?
        .filter(col("max").gt(lit(113)))?
        .with_column("sample", sample_udf.call(vec![col("max")]))?;

    ds.clone().print_physical_plan().await?;
    ds.clone().print_stream().await?;

    Ok(())
}

#[derive(Debug, Clone)]
struct SampleUdf {
    signature: Signature,
}

impl SampleUdf {
    fn new() -> Self {
        Self {
            signature: Signature::exact(vec![DataType::Float64], Volatility::Immutable),
        }
    }
}

impl ScalarUDFImpl for SampleUdf {
    fn as_any(&self) -> &dyn Any {
        self
    }

    fn name(&self) -> &str {
        "sample_udf"
    }

    fn signature(&self) -> &Signature {
        &self.signature
    }

    fn return_type(&self, _arg_types: &[DataType]) -> datafusion::error::Result<DataType> {
        Ok(DataType::Float64)
    }

    fn invoke(&self, args: &[ColumnarValue]) -> datafusion::error::Result<ColumnarValue> {
        assert_eq!(args.len(), 1);
        let value = &args[0];
        assert_eq!(value.data_type(), DataType::Float64);

        let args = ColumnarValue::values_to_arrays(args)?;
        let value = as_float64_array(&args[0]).expect("cast failed");

        let array = value
            .iter()
            .map(|v| match v {
                Some(f) => {
                    let value = f + 20_f64;
                    Some(value)
                }
                _ => None,
            })
            .collect::<Float64Array>();

        Ok(ColumnarValue::from(Arc::new(array) as ArrayRef))
    }

    fn output_ordering(
        &self,
        input: &[ExprProperties],
    ) -> datafusion::error::Result<SortProperties> {
        Ok(input[0].sort_properties)
    }
}

File: py-denormalized/python/examples/udf_example.py
"""stream_aggregate example."""

import json
import signal
import sys

import pyarrow as pa
import pyarrow.compute as pc
from denormalized import Context
from denormalized.datafusion import col
from denormalized.datafusion import functions as f
from denormalized.datafusion import lit, udf


def signal_handler(sig, frame):
    sys.exit(0)


signal.signal(signal.SIGINT, signal_handler)

bootstrap_server = "localhost:9092"
timestamp_column = "occurred_at_ms"

sample_event = {
    "occurred_at_ms": 100,
    "sensor_name": "foo",
    "reading": 0.0,
}


def gt(lhs: pa.Array, rhs: pa.Scalar) -> pa.Array:
    return pc.greater(lhs, rhs)


greater_than_udf = udf(gt, [pa.float64(), pa.float64()], pa.bool_(), "stable")


def print_batch(rb: pa.RecordBatch):
    if not len(rb):
        return
    print(rb)


ds = Context().from_topic(
    "temperature",
    json.dumps(sample_event),
    bootstrap_server,
    timestamp_column,
)

ds.window(
    [col("sensor_name")],
    [
        f.count(col("reading"), distinct=False, filter=None).alias("count"),
        f.min(col("reading")).alias("min"),
        f.max(col("reading")).alias("max"),
        f.avg(col("reading")).alias("average"),
    ],
    1000,
    None,
).with_column(
    "greater_than",
    greater_than_udf(
        col("count"),
        lit(1400.0),
    ),
).sink(
    print_batch
)
